File: solutions/2026-01-22/python/2026-01-22-minimum-pair-removal-to-sort-array-i.py
```python
# Summary: Repeatedly merge adjacent pairs with the minimum sum until the array is non-decreasing.
# Link: https://leetcode.com/problems/minimum-pair-removal-to-sort-array-i/
# Approach:
# We can simulate the process. In each step, we need to find the adjacent pair with the minimum sum.
# If there are multiple pairs with the same minimum sum, we choose the leftmost one.
# We then replace this pair with their sum. We repeat this until the array is sorted in non-decreasing order.
#
# To efficiently find the minimum sum pair, we can iterate through the adjacent pairs in each step.
# A `while` loop will continue as long as the array is not sorted.
# Inside the loop, we find the minimum sum and its index.
# Then, we update the array by replacing the pair with their sum.
# The count of operations is incremented in each iteration of the `while` loop.
#
# Time Complexity:
# In the worst case, the array length can be up to 50. Each operation reduces the array length by 1.
# So, there can be at most N-1 operations where N is the initial length of the array.
# In each operation, we iterate through the array to find the minimum sum pair. This takes O(N) time.
# Therefore, the overall time complexity is O(N^2) in the worst case, where N is the initial length of nums.
# Given N <= 50, N^2 is at most 2500, which is acceptable.
#
# Space Complexity:
# We are modifying the input array in-place. We are using a few variables to store the minimum sum, index, etc.
# This uses a constant amount of extra space.
# Therefore, the space complexity is O(1).

class Solution:
    def minOperations(self, nums: list[int]) -> int:
        """
        Calculates the minimum number of operations to make the array non-decreasing
        by repeatedly merging the adjacent pair with the minimum sum.
        """
        operations = 0  # Initialize the count of operations

        # Continue operations until the array is sorted non-decreasingly
        while not self.is_non_decreasing(nums):
            min_sum = float('inf')  # Initialize minimum sum to infinity
            min_idx = -1  # Initialize index of the pair with minimum sum

            # Iterate through adjacent pairs to find the one with the minimum sum
            # We only need to check up to len(nums) - 2 because we are looking at pairs (nums[i], nums[i+1])
            for i in range(len(nums) - 1):
                current_sum = nums[i] + nums[i+1]
                # If the current sum is smaller than the minimum sum found so far,
                # or if the current sum is equal to the minimum sum but it's a
                # leftmost pair (smaller index), update min_sum and min_idx.
                if current_sum < min_sum:
                    min_sum = current_sum
                    min_idx = i

            # If a valid pair was found (i.e., the array was not already sorted and had at least 2 elements)
            if min_idx != -1:
                # Replace the pair (nums[min_idx], nums[min_idx+1]) with their sum
                # The new element will be at index min_idx.
                # We can achieve this by replacing nums[min_idx] with the sum and then removing nums[min_idx+1].
                nums[min_idx] = nums[min_idx] + nums[min_idx+1]
                nums.pop(min_idx + 1)
                operations += 1  # Increment the operation count
            else:
                # This case should ideally not be reached if is_non_decreasing works correctly,
                # but as a safeguard, break if no pair is found (e.g., array of length 1).
                break

        return operations

    def is_non_decreasing(self, arr: list[int]) -> bool:
        """
        Checks if an array is non-decreasing.
        An array is non-decreasing if each element is greater than or equal to its previous element.
        """
        # Iterate from the second element to the end of the array
        for i in range(1, len(arr)):
            # If any element is smaller than its previous element, the array is not non-decreasing
            if arr[i] < arr[i-1]:
                return False
        # If the loop completes without finding any violation, the array is non-decreasing
        return True

```

File: solutions/2026-01-22/java/2026-01-22-minimum-pair-removal-to-sort-array-i.java
```java
// Problem: Minimum Pair Removal to Sort Array I
// Link: https://leetcode.com/problems/minimum-pair-removal-to-sort-array-i/
//
// Summary: Repeatedly merge the adjacent pair with the minimum sum until the array is non-decreasing.
//
// Approach:
// The problem requires us to repeatedly find the adjacent pair with the minimum sum and replace it with their sum.
// This process continues until the array becomes non-decreasing.
// Since the array length is small (<= 50), a simulation-based approach is feasible.
// We can use a loop that continues as long as the array is not sorted.
// Inside the loop:
// 1. Find the index of the adjacent pair with the minimum sum.
//    We iterate through all adjacent pairs and keep track of the minimum sum found and the index of its left element.
//    If multiple pairs have the same minimum sum, we choose the leftmost one.
// 2. If no merge is needed (i.e., the array is already sorted), break the loop.
// 3. If a merge is performed:
//    - Replace the element at the identified index with the sum of the pair.
//    - Remove the element at the next index.
// This simulation directly models the problem statement.
//
// Time Complexity:
// In the worst case, we might need to perform N-1 merges (where N is the initial length of the array) to reduce it to a single element.
// In each merge operation:
// - Finding the minimum sum pair takes O(L) time, where L is the current length of the list.
// - Modifying the list (e.g., using ArrayList.remove and ArrayList.set) takes O(L) time in the worst case due to element shifting.
// Since L decreases from N to 1, the total time complexity is roughly O(N^3).
// Given N <= 50, N^3 is acceptable.
//
// Space Complexity:
// We use an ArrayList to store and modify the array, which takes O(N) space.
// Other variables use constant space.
// Therefore, the space complexity is O(N).

import java.util.ArrayList;
import java.util.List;

class Solution {
    /**
     * Repeatedly merges the adjacent pair with the minimum sum until the array is non-decreasing.
     *
     * @param nums The input array of integers.
     * @return The minimum number of operations needed to make the array non-decreasing.
     */
    public int minimumOperations(int[] nums) {
        // Convert the input array to a List for easier modification (addition/removal)
        List<Integer> numList = new ArrayList<>();
        for (int num : nums) {
            numList.add(num);
        }

        int operations = 0;

        // Continue performing operations as long as the list is not sorted non-decreasingly
        while (!isNonDecreasing(numList)) {
            // Find the adjacent pair with the minimum sum
            int minSum = Integer.MAX_VALUE;
            int minIndex = -1; // Index of the left element of the pair with minimum sum

            // Iterate through all adjacent pairs
            for (int i = 0; i < numList.size() - 1; i++) {
                int currentSum = numList.get(i) + numList.get(i + 1);
                // If the current sum is less than the minimum sum found so far,
                // update minSum and minIndex.
                // If sums are equal, we prefer the leftmost pair, so we only update if currentSum < minSum.
                if (currentSum < minSum) {
                    minSum = currentSum;
                    minIndex = i;
                }
            }

            // If minIndex is still -1, it means the list has only one element or is already sorted.
            // However, the outer while loop condition handles the already sorted case.
            // If the list has only one element, the loop condition will also handle it.
            // So, minIndex will always be valid if the loop continues.

            // Perform the merge operation
            // Get the elements of the pair to be merged
            int firstElement = numList.get(minIndex);
            int secondElement = numList.get(minIndex + 1);

            // Replace the first element with their sum
            numList.set(minIndex, firstElement + secondElement);
            // Remove the second element of the pair
            numList.remove(minIndex + 1);

            // Increment the operation count
            operations++;
        }

        // Return the total number of operations performed
        return operations;
    }

    /**
     * Checks if the given list of integers is non-decreasing.
     *
     * @param list The list to check.
     * @return true if the list is non-decreasing, false otherwise.
     */
    private boolean isNonDecreasing(List<Integer> list) {
        // An empty list or a list with one element is considered non-decreasing.
        if (list.size() <= 1) {
            return true;
        }
        // Iterate through the list and check if any element is smaller than its predecessor.
        for (int i = 0; i < list.size() - 1; i++) {
            if (list.get(i) > list.get(i + 1)) {
                return false; // Found a violation of the non-decreasing property
            }
        }
        // If the loop completes without finding any violations, the list is non-decreasing.
        return true;
    }
}
```